packages/babel-helpers/src/index.js

```diff
@@ -1,10 +1,208 @@
+
+      if (binding && binding.scope.path.isProgram()) {
+        exportBindingAssignments.push(makePath(child));
+      }
+    },
+  });
+
+  if (!exportPath) throw new Error("Helpers must default-export something.");
+
+  // Process these in reverse so that mutating the references does not invalidate any later paths in
+  // the list.
+  exportBindingAssignments.reverse();
+
+  return {
+    globals: Array.from(globals),
+    localBindingNames: Array.from(localBindingNames),
+    exportBindingAssignments,
+    exportPath,
+    exportName,
+  };
+}
+
+/**
+ * Given a helper AST and information about how it will be used, update the AST to match the usage.
+ */
+function permuteHelperAST(file, metadata, id, localBindings) {
+  if (localBindings && !id) {
+    throw new Error("Unexpected local bindings for module-based helpers.");
+  }
+
+  if (!id) return;
+
+  const {
+    localBindingNames,
+    exportBindingAssignments,
+    exportPath,
+    exportName,
+  } = metadata;
+
+  const toRename = {};
+  const bindings = new Set(localBindings || []);
+  localBindingNames.forEach(name => {
+    let newName = name;
+    while (bindings.has(newName)) newName = "_" + newName;
+
+    if (newName !== name) toRename[name] = newName;
+  });
+
+  if (id.type === "Identifier" && exportName !== id.name) {
+    toRename[exportName] = id.name;
+  }
+
+  traverse(file, {
+    Program(path) {
+      const exp = path.get(exportPath);
+      const decl = exp.get("declaration");
+      if (id.type === "Identifier") {
+        if (decl.isFunctionDeclaration()) {
+          exp.replaceWith(decl);
+        } else {
+          exp.replaceWith(
+            t.variableDeclaration("var", [t.variableDeclarator(id, decl.node)]),
+          );
+        }
+      } else if (id.type === "MemberExpression") {
+        if (decl.isFunctionDeclaration()) {
+          exportBindingAssignments.forEach(assignPath => {
+            const assign = path.get(assignPath);
+            assign.replaceWith(t.assignmentExpression("=", id, assign.node));
+          });
+          exp.replaceWith(decl);
+          path.pushContainer(
+            "body",
+            t.assignmentExpression("=", id, t.identifier(exportName)),
+          );
+        } else {
+          exp.replaceWith(t.assignmentExpression("=", id, decl.node));
+        }
+      } else {
+        throw new Error("Unexpected helper format.");
+      }
+
+      Object.keys(toRename).forEach(name => {
+        path.scope.rename(name, toRename[name]);
+      });
+
+      // We only use "traverse" for all the handy scoping helpers, so we can stop immediately without
+      // actually doing the traversal.
+      path.stop();
+    },
+  });
+}
+
+const helperData = {};
+function loadHelper(name) {
+  if (!helperData[name]) {
+    const fn = () => {
+      const ast = helpers[name]();
+      return t.file(t.program(Array.isArray(ast) ? ast : [ast]));
+    };
+    if (!fn) throw new ReferenceError(`Unknown helper ${name}`);
```

Comment:
I don't understand this. It's always a function, you just defined it.

Reply:
Yeah I 💯  put this in the wrong place, not sure what I was thinking.